src/stories/TextInput.tsx

```diff
@@ -0,0 +1,161 @@
+//on render, component needs the cursor position
+
+import React, { useRef, useState, useEffect } from "react";
+import InfoIcon from "@mui/icons-material/Info";
+import "./textinput.css";
+import TitleIcon from "@mui/icons-material/Title";
+import KeyboardReturnIcon from "@mui/icons-material/KeyboardReturn";
+
+interface TextInputProps {
+  placeholder: string;
+  isDisabled: boolean;
+  isPassword: boolean;
+  validationRegex: RegExp | string;
+}
+
+const TextInput = ({
+  placeholder,
+  isDisabled,
+  isPassword,
+  validationRegex,
+}: TextInputProps) => {
+  const [infoHover, setInfoHover] = useState(false);
+  const [infoMouseDown, setInfoMouseDown] = useState(false);
+  const [submitHover, setSubmitHover] = useState(false);
+  const [submitMouseDown, setSubmitMouseDown] = useState(false);
+  const [selectionStart, setSelectionStart] = useState(0);
+  const [selectionEnd, setSelectionEnd] = useState(0);
+
+  const [realValue, setRealValue] = useState("");
+  const [displayValue, setDisplayValue] = useState("");
+
+  const textareaRef = useRef<HTMLTextAreaElement>(null);
+
+  const regex = new RegExp(validationRegex);
+  const isValid = regex.test(realValue);
+  const validCharacters = /^[^\s\x00-\x1F\x7F-\x9F\u2028\u2029]$/;
+
+  const handleKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
+    if (e.key === "Backspace") {
+      if (selectionStart === selectionEnd) {
+        setRealValue(
+          realValue.slice(0, selectionStart - 1) +
+            realValue.slice(selectionEnd, realValue.length)
+        );
+      } else {
+        setRealValue(
+          realValue.slice(0, selectionStart) +
+            realValue.slice(selectionEnd, realValue.length)
+        );
+      }
+      if (textareaRef.current) {
+        textareaRef.current.setSelectionRange(0, 0);
+      }
+      //flickers
+    } else if (validCharacters.test(e.key)) {
+      setRealValue(
+        realValue.slice(0, selectionStart) +
+          e.key +
+          realValue.slice(selectionEnd, realValue.length)
+      );
+    }
```

Comment:
Move the code out of the useEffect to the bottom here, and then create a variable that stores realValue in the context of this function instead of in state. There's no need to store realValue in state.

setDisplayValue(isPassword ? "•".repeat(realValue.length) : realValue);